File: ios/VoteGlobal/Presentation/Governance/GovernanceViewModel.swift
import Foundation
import Combine

// MARK: - Models

struct GovernanceProposal: Identifiable, Decodable {
    let proposalId: String
    let proposer: String
    let description: String
    let state: String
    let voteStart: Int
    let voteEnd: Int
    let votes: ProposalVotes

    var id: String { proposalId }

    var stateColor: String {
        switch state {
        case "Active":    return "statusOpen"
        case "Succeeded": return "statusFinalised"
        case "Defeated":  return "statusClosed"
        case "Executed":  return "statusFinalised"
        default:          return "statusTallying"
        }
    }
}

struct ProposalVotes: Decodable {
    let `for`: String
    let against: String
    let abstain: String

    var forDouble:     Double { Double(`for`)     ?? 0 }
    var againstDouble: Double { Double(against)   ?? 0 }
    var abstainDouble: Double { Double(abstain)   ?? 0 }
    var total:         Double { forDouble + againstDouble + abstainDouble }
}

// MARK: - ViewModel

@MainActor
final class GovernanceViewModel: ObservableObject {

    enum State {
        case idle, loading, loaded([GovernanceProposal]), error(String)
    }

    enum VoteSupport: Int { case against = 0, `for` = 1, abstain = 2 }

    @Published var state: State = .idle
    @Published var castingVote = false
    @Published var voteResult: String? = nil

    private let baseURL: String
    private let session = URLSession.shared

    init(baseURL: String = ProcessInfo.processInfo.environment["API_BASE_URL"] ?? "http://localhost:3000") {
        self.baseURL = baseURL
    }

    func loadProposals() async {
        state = .loading
        do {
            var req = URLRequest(url: URL(string: "\(baseURL)/api/v1/governance/proposals")!)
            req.timeoutInterval = 15
            let (data, _) = try await session.data(for: req)
            let decoded = try JSONDecoder().decode(ProposalsResponse.self, from: data)
            state = .loaded(decoded.proposals)
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    func castVote(proposalId: String, support: VoteSupport) async {
        castingVote = true
        voteResult  = nil
        defer { castingVote = false }

        do {
            var req = URLRequest(url: URL(string: "\(baseURL)/api/v1/governance/proposals/\(proposalId)/vote")!)
            req.httpMethod = "POST"
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONEncoder().encode(["support": support.rawValue])
            let (data, response) = try await session.data(for: req)

            guard (response as? HTTPURLResponse)?.statusCode == 200 else {
                let err = try? JSONDecoder().decode(APIError.self, from: data)
                throw URLError(.badServerResponse)
            }
            let result = try JSONDecoder().decode(TxResponse.self, from: data)
            voteResult = "TX: \(result.txHash)"
            await loadProposals() // refresh
        } catch {
            voteResult = "Vote failed: \(error.localizedDescription)"
        }
    }

    private struct ProposalsResponse: Decodable {
        let proposals: [GovernanceProposal]
    }
    private struct TxResponse: Decodable { let txHash: String }
    private struct APIError: Decodable   { let error: String }
}

File: android/app/src/main/kotlin/com/TheVoteApp/presentation/governance/GovernanceViewModel.kt
package com.TheVoteApp.presentation.governance

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

// ── Data models ──────────────────────────────────────────────────────────────

@Serializable
data class GovernanceProposal(
    val proposalId: String,
    val proposer: String,
    val description: String,
    val state: String,
    val voteStart: Int,
    val voteEnd: Int,
    val votes: ProposalVotes,
)

@Serializable
data class ProposalVotes(
    val `for`: String,
    val against: String,
    val abstain: String,
) {
    val forDouble:     Double get() = `for`.toDoubleOrNull()     ?: 0.0
    val againstDouble: Double get() = against.toDoubleOrNull()   ?: 0.0
    val abstainDouble: Double get() = abstain.toDoubleOrNull()   ?: 0.0
    val total:         Double get() = forDouble + againstDouble + abstainDouble
}

@Serializable
private data class ProposalsResponse(val proposals: List<GovernanceProposal>)

@Serializable
private data class TxResponse(val txHash: String)

// Support values matching OZ Governor: 0=Against, 1=For, 2=Abstain
enum class VoteSupport(val value: Int, val label: String) {
    FOR(1, "For"), AGAINST(0, "Against"), ABSTAIN(2, "Abstain")
}

// ── UI State ─────────────────────────────────────────────────────────────────

sealed interface GovernanceUiState {
    data object Loading : GovernanceUiState
    data class  Loaded(val proposals: List<GovernanceProposal>) : GovernanceUiState
    data class  Error(val message: String) : GovernanceUiState
}

// ── ViewModel ────────────────────────────────────────────────────────────────

class GovernanceViewModel(
    private val okHttpClient: OkHttpClient = OkHttpClient(),
    private val baseUrl: String = "http://10.0.2.2:3000",  // Android emulator localhost
) : ViewModel() {

    private val _uiState = MutableStateFlow<GovernanceUiState>(GovernanceUiState.Loading)
    val uiState: StateFlow<GovernanceUiState> = _uiState.asStateFlow()

    private val _castingVote = MutableStateFlow(false)
    val castingVote: StateFlow<Boolean> = _castingVote.asStateFlow()

    private val _voteResult = MutableStateFlow<String?>(null)
    val voteResult: StateFlow<String?> = _voteResult.asStateFlow()

    private val json = Json { ignoreUnknownKeys = true }

    init { loadProposals() }

    fun loadProposals() {
        viewModelScope.launch {
            _uiState.value = GovernanceUiState.Loading
            try {
                val request = Request.Builder()
                    .url("$baseUrl/api/v1/governance/proposals")
                    .build()
                val body = okHttpClient.newCall(request).execute().use { it.body!!.string() }
                val response = json.decodeFromString<ProposalsResponse>(body)
                _uiState.value = GovernanceUiState.Loaded(response.proposals)
            } catch (e: Exception) {
                _uiState.value = GovernanceUiState.Error(e.message ?: "Unknown error")
            }
        }
    }

    fun castVote(proposalId: String, support: VoteSupport) {
        viewModelScope.launch {
            _castingVote.value = true
            _voteResult.value  = null
            try {
                val body = """{"support":${support.value}}"""
                    .toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url("$baseUrl/api/v1/governance/proposals/$proposalId/vote")
                    .post(body)
                    .build()
                val responseBody = okHttpClient.newCall(request).execute().use { it.body!!.string() }
                val tx = json.decodeFromString<TxResponse>(responseBody)
                _voteResult.value = "TX: ${tx.txHash.take(18)}…"
                loadProposals()
            } catch (e: Exception) {
                _voteResult.value = "Vote failed: ${e.message}"
            } finally {
                _castingVote.value = false
            }
        }
    }
}
